ceph_fs.h: add separate owner_{u,g}id fields

mihalicyn opened the pull request:
This patch adds separate fields to pass inode owner's UID/GID for operations which create new inodes:
CEPH_MDS_OP_CREATE, CEPH_MDS_OP_MKNOD, CEPH_MDS_OP_MKDIR, CEPH_MDS_OP_SYMLINK.

Before we have used caller_{u,g}id fields for two purposes:
- for MDS permission checks
- to set UID/GID for new inodes

but during a long discussion around adding idmapped mounts support to cephfs kernel client we decided to extend wire protocol [1].

Originally, Christian Brauner pointed to this problem [2], but a few initial versions of "ceph: support idmapped mounts" patchset were based on a compromise between bringing idmapped mounts support and changing on-wire data structures. The only problem with this approach is that if we have UID/GID-based permission checks on the MDS side they won't work as intended. Xiubo Li pointed to this issue and said that it's critical enough to not being ignored.

This patch does not change behavior for client or MDS, because in the current client implementation we do:
```
req->set_inode_owner_uid_gid(perms.uid(), perms.gid());
```
so, these new fields contain the same value as caller_{u,g}id.

This new extension is used in a non-trivial way in the Linux kernel client:
https://github.com/mihalicyn/linux/commits/fs.idmapped.ceph

In addition to new fields we also add CEPHFS_FEATURE_HAS_OWNER_UIDGID feature bit. It can be used on the client-side
to check if MDS supports these new fields and can handle them properly.

Fixes: https://tracker.ceph.com/issues/62217

[1] https://lore.kernel.org/lkml/CAEivzxcipdTQ9-b2zk-7-AMDfwPk2Brtp=X6H8xXsHMEtQJKFQ@mail.gmail.com/
[2] https://lore.kernel.org/all/20220104140414.155198-3-brauner@kernel.org/


cc @lxbsz

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

mihalicyn commented:
jenkins test make check

lxbsz commented on src/messages/MClientRequest.h:
```diff
@@ -161,6 +161,45 @@ class MClientRequest final : public MMDSOp {
+      break;
+    default:
+      ceph_assert(false);
+      return std::make_pair(uid, gid);
```
Please remove this line.

mihalicyn replied:
done, thanks!

lxbsz commented on src/include/ceph_fs.h:
```diff
@@ -587,10 +587,18 @@ union ceph_mds_request_args {
+	} __attribute__ ((packed)) symlink;
+	struct {
+		__le32 owner_uid, owner_gid;
+	} __attribute__ ((packed)) rename;
```
Why not just define this in ` struct ceph_mds_request_head{}` directly just after `caller_{uid,gid}` ? And set them to `-1` as default and only in the `WRITE` requests will they make sense ?

And also please be careful when extending the members in ` struct ceph_mds_request_args{}`, because the kclient will just copy the memories and will always assume the size of the memories are the same between kclient and MDS.

mihalicyn replied:
My motivation to extend `union ceph_mds_request_args` was to save some bytes :) By adding these two fields to the union I'm not increasing it's size, because a biggest member of union is `setattr` and I'm not extending it.

If you think that it's still better and safer to extend `struct ceph_mds_request_head` I'll do. Just confirm, please.

lxbsz replied:
> My motivation to extend `union ceph_mds_request_args` was to save some bytes in network packets :) By adding these two fields to the union I'm not increasing it's size, because a biggest member of union is `setattr` and I'm not extending it.
> 
> Second reason it logical. I'm adding new fields only to those union members, who need them.
> 
> If you think that it's still better and safer to extend `struct ceph_mds_request_head` I'll do. Just confirm, please.

Yeah, please. Or it will be hard to maintain in future.

vshankar replied:
I'm with @lxbsz on this one.

mihalicyn replied:
Have done. Thanks!

lxbsz commented on src/messages/MClientRequest.h:
```diff
@@ -161,6 +161,45 @@ class MClientRequest final : public MMDSOp {
+    unsigned uid = 0;
+    unsigned gid = 0;
+
+    if (head.version < 3)
```
This is incorrect. You should do the version check when `encoding/decoding` the `head` message in https://github.com/ceph/ceph/blob/main/src/include/ceph_fs.h#L659-L707C1.

mihalicyn replied:
Yes, I've seen that there is some version checking logic in encoding/decoding, but at the same time in this code we do nothing with `h.args` (`bl.append((char*)&h.args, sizeof(h.args));`), but only with handling direct members of `struct ceph_mds_request_head`.

At the same time we have this in the current code:
https://github.com/ceph/ceph/blob/8c4d93c2d72698060289579a704fac9bae07da41/src/messages/MClientRequest.h#L130

I'm ready to rework it as you propose, but please confirm.

lxbsz replied:
Please fix this together with https://github.com/ceph/ceph/pull/52575#discussion_r1271622835.

mihalicyn replied:
Have done!

lxbsz commented on src/messages/MClientRequest.h:
```diff
@@ -292,6 +331,15 @@ class MClientRequest final : public MMDSOp {
+      head.op == CEPH_MDS_OP_SYMLINK) {
+      auto [owner_uid, owner_gid] = get_owner_uid_gid();
+      out << " inode_owner_uid=" << owner_uid
+	  << ", inode_owner_gid=" << owner_gid;
```
`s/inode_owner_uid/owner_uid/`
`s/inode_owner_gid/owner_gid/`

mihalicyn replied:
done

lxbsz commented on src/mds/Server.cc:
```diff
@@ -3506,10 +3508,10 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
   }
 
-  _inode->uid = mdr->client_request->get_caller_uid();
+  _inode->uid = owner_uid;
```
Please define two similar helpers `get_owner_uid()/get_owner_gid()` instead.

mihalicyn replied:
have done.

lxbsz commented on src/mds/Server.cc:
```diff
@@ -3506,10 +3508,10 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
     }
   } else {
-    _inode->gid = mdr->client_request->get_caller_gid();
+    _inode->gid = owner_gid;
```
What will happen when the requests from an old kclient ?

mihalicyn replied:
With recent pull request revision, it will take `caller_{u,g}id` values to fill `owner_{u,g}id` fields.

lxbsz commented:
@mihalicyn Could you raise on ceph tracker and add the detail examples, which as you did in the kernel mail list, there ? This could help use to understand it easier ?

mihalicyn commented:
> @mihalicyn Could you raise on ceph tracker and add the detail examples, which as you did in the kernel mail list, there ? This could help use to understand it easier ?

Sure. Will do!

vshankar commented:
> This new extension will be used in a non-trivial way in the Linux kernel client.
> 
> [1] https://lore.kernel.org/lkml/CAEivzxcipdTQ9-b2zk-7-AMDfwPk2Brtp=X6H8xXsHMEtQJKFQ@mail.gmail.com/ [2] https://lore.kernel.org/all/20220104140414.155198-3-brauner@kernel.org/

Thanks for the pointer @mihalicyn - it helps to speed up review since I wasn't following the discussion closely on the kernel list.

mihalicyn commented:
I've started reworking kernel client code to support this protocol change and noticed that kernel client still not supports feature from https://github.com/ceph/ceph/pull/45669

Xiubo, do you have any plans to support it? I can try to support it by myself if you don't mind by using inspiration from https://github.com/ceph/ceph/commit/cbd7e3040208a5d89b9a95d4161dbd908da51079 :)

As far as I understand we *have* to support `CEPHFS_FEATURE_32BITS_RETRY_FWD` feature before `CEPHFS_FEATURE_HAS_OWNER_UIDGID`. Please correct me if I'm wrong.

lxbsz commented on src/client/MetaRequest.h:
```diff
@@ -153,6 +153,20 @@ struct MetaRequest {
+		head.op == CEPH_MDS_OP_RENAME     ||
+		head.op == CEPH_MDS_OP_MKSNAP     ||
+		head.op == CEPH_MDS_OP_MKDIR      ||
+		head.op == CEPH_MDS_OP_SYMLINK);
```
Maybe here we can reuse the `bool Client::is_dir_operation(MetaRequest *req)`.

mihalicyn replied:
hm, we can but not precisely. Set of operations in `is_dir_operation` is not the same as here. Technically we can remove this assertion at all, but I've decided to add it just as precaution.

mihalicyn replied:
We can also use something like `op & CEPH_MDS_OP_WRITE`, it's close but not the same precisely too.

lxbsz replied:
Yeah, then please define a similar macro for this. Then we can reuse it in both.

mihalicyn replied:
sure! Will do.

mihalicyn replied:
Fixed.

lxbsz commented on src/include/ceph_fs.h:
```diff
@@ -654,6 +654,8 @@ struct ceph_mds_request_head {
 	__le32 ext_num_retry;          /* new count retry attempts */
 	__le32 ext_num_fwd;            /* new count fwd attempts */
+
+	__le32 owner_uid, owner_gid;   /* used for OPs which create inodes */
```
We'd better initialize them to `-1` because `0` is still a validate value here.

mihalicyn replied:
Have done

mihalicyn commented:
Offtop:

is it okay that `../src/vstart.sh -d -n -l` fails to work with this:
```
/home/alex/storage/dev/ceph/build/bin/ceph -c /home/alex/storage/dev/ceph/build/ceph.conf -k /home/alex/storage/dev/ceph/build/keyring fs volume ls 
no valid command found; 10 closest matches:
fs snapshot mirror enable [<fs_name>]
fs snapshot mirror disable [<fs_name>]
fs snapshot mirror peer_add <fs_name> [<remote_cluster_spec>] [<remote_fs_name>] [<remote_mon_host>] [<cephx_key>]
fs snapshot mirror peer_list [<fs_name>]
fs snapshot mirror peer_remove <fs_name> [<peer_uuid>]
fs snapshot mirror peer_bootstrap create <fs_name> <client_name> [<site_name>]
fs snapshot mirror peer_bootstrap import <fs_name> [<token>]
fs snapshot mirror add <fs_name> [<path>]
fs snapshot mirror remove <fs_name> [<path>]
fs snapshot mirror dirmap <fs_name> [<path>]
Error EINVAL: invalid command
```

As far as I understand this command fails (from `vstart.sh` source):
```
	    # wait for volume module to load
	    while ! ceph_adm fs volume ls ; do sleep 1 ; done # <<<<<<<< HERE
            local fs=0
            for name in a b c d e f g h i j k l m n o p
            do
                ceph_adm fs volume create ${name}
                ceph_adm fs authorize ${name} "client.fs_${name}" / rwp >> "$keyring_fn"
                fs=$(($fs + 1))
                [ $fs -eq $CEPH_NUM_FS ] && break
            done
        fi
```

`ceph_adm` syntax was changed recently, but `vstart.sh` wasn't updated?

lxbsz commented:
> 



> Offtop:
> 
> is it okay that `../src/vstart.sh -d -n -l` fails to work with this:
> 
> ```
> /home/alex/storage/dev/ceph/build/bin/ceph -c /home/alex/storage/dev/ceph/build/ceph.conf -k /home/alex/storage/dev/ceph/build/keyring fs volume ls 
> no valid command found; 10 closest matches:
> fs snapshot mirror enable [<fs_name>]
> fs snapshot mirror disable [<fs_name>]
> fs snapshot mirror peer_add <fs_name> [<remote_cluster_spec>] [<remote_fs_name>] [<remote_mon_host>] [<cephx_key>]
> fs snapshot mirror peer_list [<fs_name>]
> fs snapshot mirror peer_remove <fs_name> [<peer_uuid>]
> fs snapshot mirror peer_bootstrap create <fs_name> <client_name> [<site_name>]
> fs snapshot mirror peer_bootstrap import <fs_name> [<token>]
> fs snapshot mirror add <fs_name> [<path>]
> fs snapshot mirror remove <fs_name> [<path>]
> fs snapshot mirror dirmap <fs_name> [<path>]
> Error EINVAL: invalid command
> ```
> 
> As far as I understand this command fails (from `vstart.sh` source):
> 
> ```
> 	    # wait for volume module to load
> 	    while ! ceph_adm fs volume ls ; do sleep 1 ; done # <<<<<<<< HERE
>             local fs=0
>             for name in a b c d e f g h i j k l m n o p
>             do
>                 ceph_adm fs volume create ${name}
>                 ceph_adm fs authorize ${name} "client.fs_${name}" / rwp >> "$keyring_fn"
>                 fs=$(($fs + 1))
>                 [ $fs -eq $CEPH_NUM_FS ] && break
>             done
>         fi
> ```
> 
> `ceph_adm` syntax was changed recently, but `vstart.sh` wasn't updated?
> 
> Upd: solved. Analogical case https://tracker.ceph.com/issues/41251
> 
> ```
> cd build
> cat out/*.log | grep ModuleNot
> ```
> 
> and install all python packages with `pip`.

You should run the ` ./install-deps.sh ` script to install the dependents.

The is what I usually use to set the cluster locally:
`MDS=3 MON=3 OSD=3 MGR=1 ../src/vstart.sh -n -X -G --msgr1`

mihalicyn commented:
> You should run the `./install-deps.sh` script to install the dependents.

yes, that's what I did at first, but it looks like it doesn't install enough pip packages, so `pip install cherrypy prettytable scipy pecan` did the trick.



> The is what I usually use to set the cluster locally:
> `MDS=3 MON=3 OSD=3 MGR=1 ../src/vstart.sh -n -X -G --msgr1`

Thanks! Will do the same for testing ;-)

mihalicyn commented:
Revision 3.
- addressed all review comments/suggestions
- prepared and tested kernel client changes (https://github.com/mihalicyn/linux/commits/fs.idmapped.ceph) with xfstests

I'll send kernel client patches to LKML once this PR will be merged or upon request from Xiubo.

lxbsz commented:
> Revision 3.
> 
> * addressed all review comments/suggestions
> * prepared and tested kernel client changes (https://github.com/mihalicyn/linux/commits/fs.idmapped.ceph) with xfstests
> 
> I'll send kernel client patches to LKML once this PR will be merged or upon request from Xiubo.
> 
> I'm still waiting for my account at tracker.ceph.com to be approved by administrator.

@mihalicyn Please go on to code and send the kclient patches first, we need to go through both ceph and kernel changes before we merge this PR to make sure we won't miss anything important for the idmapping feature.

Please code the kclient patches basing on the `testing` branch, I will apply my patch [1] to it.

[1] https://patchwork.kernel.org/project/ceph-devel/patch/20230725104438.386461-1-xiubli@redhat.com/

vshankar commented:
> Revision 2.
> 
> * addressed all review comments/suggestions
> * rebased to the latest `main`
> * added `CEPHFS_FEATURE_HAS_OWNER_UIDGID` it will be useful in kernel client to detect old MDS servers if needed

Could you elaborate a bit more on this? Seeing the client side changes, its the client side that's setting the owner_{g,u}id in the request and the MDS knows how to either decode from the bufferlist or set the owner_{g,u}id from the caller_{g,u}id for older clients.

(maybe I missed a discussion somewhere)

mihalicyn commented:
> > Revision 3.
> > 
> > * addressed all review comments/suggestions
> > * prepared and tested kernel client changes (https://github.com/mihalicyn/linux/commits/fs.idmapped.ceph) with xfstests
> > 
> > I'll send kernel client patches to LKML once this PR will be merged or upon request from Xiubo.
> > I'm still waiting for my account at tracker.ceph.com to be approved by administrator.
> 
> @mihalicyn Please go on to code and send the kclient patches first, we need to go through both ceph and kernel changes before we merge this PR to make sure we won't miss anything important for the idmapping feature.
> 
> Please code the kclient patches basing on the `testing` branch, I will apply my patch [1] to it.
> 
> [1] https://patchwork.kernel.org/project/ceph-devel/patch/20230725104438.386461-1-xiubli@redhat.com/

Have sent.

https://lore.kernel.org/all/20230726141026.307690-1-aleksandr.mikhalitsyn@canonical.com/#t

Thanks!

vshankar commented:
> > Revision 2.
> > 
> > * addressed all review comments/suggestions
> > * rebased to the latest `main`
> > * added `CEPHFS_FEATURE_HAS_OWNER_UIDGID` it will be useful in kernel client to detect old MDS servers if needed
> 
> Could you elaborate a bit more on this? Seeing the client side changes, its the client side that's setting the owner_{g,u}id in the request and the MDS knows how to either decode from the bufferlist or set the owner_{g,u}id from the caller_{g,u}id for older clients.
> 
> (maybe I missed a discussion somewhere)

Ok - in the kclient change you mentioned that v7 uses CEPHFS_FEATURE_HAS_OWNER_UIDGID. I'll look there. nvm

mihalicyn commented:
> > Revision 2.
> > 
> > * addressed all review comments/suggestions
> > * rebased to the latest `main`
> > * added `CEPHFS_FEATURE_HAS_OWNER_UIDGID` it will be useful in kernel client to detect old MDS servers if needed
> 
> Could you elaborate a bit more on this? Seeing the client side changes, its the client side that's setting the owner_{g,u}id in the request and the MDS knows how to either decode from the bufferlist or set the owner_{g,u}id from the caller_{g,u}id for older clients.
> 
> (maybe I missed a discussion somewhere)

Hi @vshankar

yes, you are absolutely right, but having a separate feature flag for it is still useful.

Let me explain on the example.
1. If we have an "old" client but "new" MDS then everything is okay, because the old client
won't send new fields, "new" MDS will take `caller_{g,u}id` values and fill `owner_{g,u}id` using them.
2. If we have a "new" client but "old" MDS then we have a problem. Imagine that "new" client
sets these new fields `owner_{g,u}id` and values are not equal to `caller_{g,u}id`. Old MDS
server will ignore this `owner_{g,u}id` but "new" client won't be ready for this behavior and can
be confused by that. That's why we need this feature flag to detect that MDS server is new enough
to properly support case when `owner_{g,u}id != caller_{g,u}id`. Otherwise, client should always
maintain `owner_{g,u}id == caller_{g,u}id`.

Commit for Linux kernel client that uses this feature:
https://github.com/mihalicyn/linux/commit/4981918fa0a76bd4614d7a354e8e3ee020a0a626

lxbsz commented:
> > > Revision 2.
> > > 
> > > * addressed all review comments/suggestions
> > > * rebased to the latest `main`
> > > * added `CEPHFS_FEATURE_HAS_OWNER_UIDGID` it will be useful in kernel client to detect old MDS servers if needed
> > 
> > 
> > Could you elaborate a bit more on this? Seeing the client side changes, its the client side that's setting the owner_{g,u}id in the request and the MDS knows how to either decode from the bufferlist or set the owner_{g,u}id from the caller_{g,u}id for older clients.
> > (maybe I missed a discussion somewhere)
> 
> Ok - in the kclient change you mentioned that v7 uses CEPHFS_FEATURE_HAS_OWNER_UIDGID. I'll look there. nvm

@mihalicyn Here in the `client/Client.cc` you are reusing the `CEPHFS_FEATURE_32BITS_RETRY_FWD` instead of `CEPHFS_FEATURE_HAS_OWNER_UIDGID`, so we can do the same in kclient. Then no need to add a new feature bit ?

mihalicyn commented:
> Here in the `client/Client.cc` you are reusing the `CEPHFS_FEATURE_32BITS_RETRY_FWD` instead of `CEPHFS_FEATURE_HAS_OWNER_UIDGID`, so we can do the same in kclient. Then no need to add a new feature bit ?

No, we can't reuse it in a kernel client. Here, in `client/Client.cc` we don't mention `CEPHFS_FEATURE_HAS_OWNER_UIDGID` only because we always have `owner_{g,u}id == caller_{g,u}id`. And we don't care if MDS supports `CEPHFS_FEATURE_HAS_OWNER_UIDGID` feature or not, because anyway inode owner will be set to the correct value.

But in kernel client we have `owner_{g,u}id != caller_{g,u}id` (if mount's idmapping used) and we need to ensure that MDS can properly handle this and won't ignore new `owner_{g,u}id` fields silently by taking `caller_{g,u}id` instead.

lxbsz commented:
> > Here in the `client/Client.cc` you are reusing the `CEPHFS_FEATURE_32BITS_RETRY_FWD` instead of `CEPHFS_FEATURE_HAS_OWNER_UIDGID`, so we can do the same in kclient. Then no need to add a new feature bit ?
> 
> No, we can't reuse it in a kernel client. Here, in `client/Client.cc` we don't mention `CEPHFS_FEATURE_HAS_OWNER_UIDGID` only because we always have `owner_{g,u}id == caller_{g,u}id`. And we don't care if MDS supports `CEPHFS_FEATURE_HAS_OWNER_UIDGID` feature or not, because anyway inode owner will be set to the correct value.
> 
> But in kernel client we have `owner_{g,u}id != caller_{g,u}id` (if mount's idmapping used) and we need to ensure that MDS can properly handle this and won't ignore new `owner_{g,u}id` fields silently by taking `caller_{g,u}id` instead.

Yeah, I checked the kclient patches. This feature bit makes sense.

mihalicyn commented:
I've added a ticket on CephFS tracker: https://tracker.ceph.com/issues/62217

mihalicyn commented:
@lxbsz Xiubo,
can you comment on this one https://lore.kernel.org/all/CA+enf=sFC-hiziuXoeDWnb7MoErc+b1PAncOjbM2rNyB4fzfwA@mail.gmail.com/#t ?

I'm ready to prepare v8 series with this. This solution seems like an optimal compromise
that allows users with older MDS versions to use idmapped mounts if they are *really* want it.

lxbsz commented:
> @lxbsz Xiubo, can you comment on this one https://lore.kernel.org/all/CA+enf=sFC-hiziuXoeDWnb7MoErc+b1PAncOjbM2rNyB4fzfwA@mail.gmail.com/#t ?
> 
> I'm ready to prepare v8 series with this. This solution seems like an optimal compromise that allows users with older MDS versions to use idmapped mounts if they are _really_ want it.
> 
> I've added a note to tracker: https://tracker.ceph.com/issues/62217?next_issue_id=62215&prev_issue_id=62218#note-1

Hi Alex,

We can backport this to old releases, that shouldn't be a big issue IMO. Regarding to add one `idmap_with_old_mds` mount option, which will make the APPs to be crazy, some requests will pass the cephx mds cap check and some won't for the same users.

I will leave this to @vshankar to determine it.

@vshankar Let me explain it. If we support the `idmap_with_old_mds` mount option in kclient and when it's enabled, for old cephs it will only set the `create/mknode/mkdir` requests' `caller_{uid,gid}` to the real `uid/gid`. That means for the same user some requests will pass the cephx mds cap check and some won't, for example the user could create the file but possibly couldn't read it.  

Is that acceptable ? I'd prefer to backport this PR to old cephs instead of supporting  `idmap_with_old_mds`. What do you think ?

Thanks

mihalicyn commented:
I personally happy with backporting of this feature to older releases, but I'm not sure that distros will rebuild ceph packages with this new changes. Usually distros only taking fixes to the stable packages but no features. (cc @stgraber)

> @vshankar Let me explain it. If we support the `idmap_with_old_mds` mount option in kclient and when it's enabled, for old cephs it will only set the `create/mknode/mkdir` requests' `caller_{uid,gid}` to the real `uid/gid`. That means for the same user some requests will pass the cephx mds cap check and some won't, for example the user could create the file but possibly couldn't read it.

Let me correct this.

First of all, vice versa for all requests except `create/mknode/mkdir` `caller_{uid,gid}` will be a real `uid/gid`. Let's synchronize our terminology. real `uid/gid` is the `uid/gid` of the caller, right? And it's not important if syscall was called against the file through idmapped mount or not, caller `uid/gid` will be the same. Christian (@brauner) and I explained this concept a few times. Idmapped mounts are not proxy that hooks and replaces a caller. Idmapped mounts are about shifting `uid/gid` between the "on-disk" and "mount visible" representation. It's very thin conceptual difference. But it exists.

Secondly, for this `idmap_with_old_mds` module option (not mount, because Stéphane proposed module option), we will have problems with permission checks on the MDS side **if and only if** UID/GID-based permission checks are set on the MDS side. For most workloads it's not a case, and users who will load cephfs kernel module with this parameter _will be aware_ of this limitation.

mihalicyn commented:
So, I think it's a good idea to backport this change, but if we combine it with providing our users with a choice to enable idmapped mounts with old MDS it will be an ideal solution.

vshankar commented:
> I will leave this to @vshankar to determine it.

Give me a day or two to go through the changes and the comments.

vshankar commented:
@lxbsz @mihalicyn Going over this now - will post my comment and response to queries.

vshankar commented:
> > Here in the `client/Client.cc` you are reusing the `CEPHFS_FEATURE_32BITS_RETRY_FWD` instead of `CEPHFS_FEATURE_HAS_OWNER_UIDGID`, so we can do the same in kclient. Then no need to add a new feature bit ?
> 
> No, we can't reuse it in a kernel client. Here, in `client/Client.cc` we don't mention `CEPHFS_FEATURE_HAS_OWNER_UIDGID` only because we always have `owner_{g,u}id == caller_{g,u}id`. And we don't care if MDS supports `CEPHFS_FEATURE_HAS_OWNER_UIDGID` feature or not, because anyway inode owner will be set to the correct value.
> 
> But in kernel client we have `owner_{g,u}id != caller_{g,u}id` (if mount's idmapping used) and we need to ensure that MDS can properly handle this and won't ignore new `owner_{g,u}id` fields silently by taking `caller_{g,u}id` instead.

If I understand this correctly, the feature bit is not used in Client.cc since the user-space (fuse) mount does not support idmapping (yet*)?

[*]: PR #45667

vshankar commented:
> > @lxbsz Xiubo, can you comment on this one https://lore.kernel.org/all/CA+enf=sFC-hiziuXoeDWnb7MoErc+b1PAncOjbM2rNyB4fzfwA@mail.gmail.com/#t ?
> > I'm ready to prepare v8 series with this. This solution seems like an optimal compromise that allows users with older MDS versions to use idmapped mounts if they are _really_ want it.
> > I've added a note to tracker: https://tracker.ceph.com/issues/62217?next_issue_id=62215&prev_issue_id=62218#note-1
> 
> Hi Alex,
> 
> We can backport this to old releases, that shouldn't be a big issue IMO. Regarding to add one `idmap_with_old_mds` mount option, which will make the APPs to be crazy, some requests will pass the cephx mds cap check and some won't for the same users.
> 
> I will leave this to @vshankar to determine it.
> 
> @vshankar Let me explain it. If we support the `idmap_with_old_mds` mount option in kclient and when it's enabled, for old cephs it will only set the `create/mknode/mkdir` requests' `caller_{uid,gid}` to the real `uid/gid`. That means for the same user some requests will pass the cephx mds cap check and some won't, for example the user could create the file but possibly couldn't read it.
> 
> Is that acceptable ? I'd prefer to backport this PR to old cephs instead of supporting `idmap_with_old_mds`. What do you think ?

We can backport this change, but we still need to work with ceph-mds versions before that. Do you mean we document that a certain version of ceph is required to use idmap mount correctly?

vshankar commented:
> > We can backport this change, but we still need to work with ceph-mds versions before that. Do you mean we document that a certain version of ceph is required to use idmap mount correctly?
> 
> In the kernel client patchset (https://lore.kernel.org/all/20230726141026.307690-4-aleksandr.mikhalitsyn@canonical.com/#Z31fs:ceph:mds_client.c) we have handling code for that:
> 
> ```
> @@ -3028,6 +3029,16 @@ static struct ceph_msg *create_request_message(struct ceph_mds_session *session,
>  	lhead = find_legacy_request_head(msg->front.iov_base,
>  					 session->s_con.peer_features);
>  
> +	if ((req->r_mnt_idmap != &nop_mnt_idmap) &&
> +	    !test_bit(CEPHFS_FEATURE_HAS_OWNER_UIDGID, &session->s_features)) {
> +		pr_err_ratelimited_client(cl,
> +			"idmapped mount is used and CEPHFS_FEATURE_HAS_OWNER_UIDGID"
> +			" is not supported by MDS. Fail request with -EIO.\n");
> +
> +		ret = -EIO;
> +		goto out_err;
> +	}
> ```
> 
> This handling code does not prevent idmapped mounts from being created but it prevents any IO requests on such mounts that create new directory/file/device node/symlink. (Effectively, idmapped mounts will be read-only.) We can't prevent idmapped mounts from being mounted as idmapped mounts handling code is not cephfs-local and I don't think that it's worth modifying global VFS kernel code just to handle this rare scenario ([discussion](https://lore.kernel.org/all/CAEivzxeQubvas2yPFYRRXr3BP7pp1HNM3b7C-PQQWy-0FpFKuQ@mail.gmail.com/)). It's worth mentioning that idmapped mounts is a tricky thing and ordinary user can't create them using `mount` utility. Idmapped mounts are always handled by software like LXC (https://github.com/lxc/lxc) and it's not a problem to do an appropriate version checks in the userspace.
> 
> And it was also proposed to extend this handling and allow to use idmapped mounts even when `CEPHFS_FEATURE_HAS_OWNER_UIDGID` is not available (if cephfs `idmap_with_old_mds` kernel module parameter is passed). This extends surface when idmapped mounts can be used, but of course, in this case UID/GID-based restrictions on the MDS side will be broken (if request comes through idmapped mount _and_ MDS version does not support `CEPHFS_FEATURE_HAS_OWNER_UIDGID`). And it's not a problem work real-life workloads, because we can't remember real users who interested in idmapped mounts and use UID/GID-based restrictions.
> 
> To conclude, we are not affecting _any_ workloads that exists now. Because all of this problems with UID/GID-restrictions will happen only if IO comes through an idmapped mount. If just "regular" mounts are used then everything will work like it works now independently on `CEPHFS_FEATURE_HAS_OWNER_UIDGID`.

I'm fine with this approach. @lxbsz Do you see any concerns? We can backport this change but we also need the `idmap_with_old_mds` mount option (or is it a module option that's been suggested?).

And, oh, can the option name be a _bit_ more intuitive signifying that when enabled, things _might_ not work as expected. Something like `enable_unsafe_idmap`?

lxbsz commented:
[...]
> I'm fine with this approach. @lxbsz Do you see any concerns? We can backport this change but we also need the `idmap_with_old_mds` mount option (or is it a module option that's been suggested?).
> 

The `idmap_with_old_mds` means to allow the idmap to work partially without backporting this PR.

I'd prefer as a module parameter instead of the the mount option.

> And, oh, can the option name be a _bit_ more intuitive signifying that when enabled, things _might_ not work as expected. Something like `enable_unsafe_idmap`?

Yeah, this looks much better for me too.

vshankar commented:
> [...]
> 
> > I'm fine with this approach. @lxbsz Do you see any concerns? We can backport this change but we also need the `idmap_with_old_mds` mount option (or is it a module option that's been suggested?).
> 
> The `idmap_with_old_mds` means to allow the idmap to work partially without backporting this PR.

Right - for the newer client and and older MDS case. Users can be running an older ceph-mds version and use a newer kernel client, so just by backporting this change we cannot avoid having the mount/module option functionality implemented.

> 
> I'd prefer as a module parameter instead of the the mount option.
> 
> > And, oh, can the option name be a _bit_ more intuitive signifying that when enabled, things _might_ not work as expected. Something like `enable_unsafe_idmap`?
> 
> Yeah, this looks much better for me too.

{enable,allow}_unsafe_idmap -- anything that makes users go o_O

mihalicyn commented:
Yes, yes module parameter is definitely better. It will be easier to deprecate module parameter after a 1-2 years, than mount option. I'm fine with `enable_unsafe_idmap`. :-)

So, then I'm starting to implement this and will send a new version of kernel client patch series.

mihalicyn commented:
It looks like even after https://github.com/ceph/ceph/commit/cbd7e3040208a5d89b9a95d4161dbd908da51079 we still need to handle each `CEPH_MDS_REQUEST_HEAD_VERSION` update separately. I've just tested old MDS (without extra fields in `struct ceph_mds_request_head`) with new kernel client and it fails:
```
2023-08-03T13:15:40.871+0200 7fe64ef5e640 10 mds.c ms_handle_accept v1:192.168.2.136:0/49354629 con 0x563962206880 session 0x563967054000
2023-08-03T13:15:40.871+0200 7fe650f62640 -1 failed to decode message of type 24 v6: End of buffer [buffer:2]
2023-08-03T13:15:40.871+0200 7fe650f62640  1 dump: 
00000000  03 00 01 00 00 00 00 00  00 00 10 00 00 00 00 00  |................|
00000010  00 00 00 00 00 00 01 01  00 00 00 00 00 00 00 00  |................|
00000020  00 00 00 00 00 00 00 00  00 00 01 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000070  00 00 01 01 00 00 00 00  00 00 00 00 00 00 00 01  |................|
00000080  00 00 00 00 00 00 00 00  00 00 00 00 5b 8c cb 64  |............[..d|
00000090  64 78 11 13 01 00 00 00  00 00 00 00 00 00 00 00  |dx..............|
000000a0  00 00 00 00 00 00 00 00  00 00 00 00              |............|
000000ac
```

So, I'll fix kernel client to send an old `struct ceph_mds_request_head` if MDS does not support `CEPHFS_FEATURE_HAS_OWNER_UIDGID`. It means that we need to update userspace client code too,
and add some version checks to it to prevent any possible breakage.

lxbsz commented:
> It looks like even after [cbd7e30](https://github.com/ceph/ceph/commit/cbd7e3040208a5d89b9a95d4161dbd908da51079) we still need to handle each `CEPH_MDS_REQUEST_HEAD_VERSION` update separately. I've just tested old MDS (without extra fields in `struct ceph_mds_request_head`) with new kernel client and it fails:
> 
> ```
> 2023-08-03T13:15:40.871+0200 7fe64ef5e640 10 mds.c ms_handle_accept v1:192.168.2.136:0/49354629 con 0x563962206880 session 0x563967054000
> 2023-08-03T13:15:40.871+0200 7fe650f62640 -1 failed to decode message of type 24 v6: End of buffer [buffer:2]
> 2023-08-03T13:15:40.871+0200 7fe650f62640  1 dump: 
> 00000000  03 00 01 00 00 00 00 00  00 00 10 00 00 00 00 00  |................|
> 00000010  00 00 00 00 00 00 01 01  00 00 00 00 00 00 00 00  |................|
> 00000020  00 00 00 00 00 00 00 00  00 00 01 00 00 00 00 00  |................|
> 00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
> *
> 00000070  00 00 01 01 00 00 00 00  00 00 00 00 00 00 00 01  |................|
> 00000080  00 00 00 00 00 00 00 00  00 00 00 00 5b 8c cb 64  |............[..d|
> 00000090  64 78 11 13 01 00 00 00  00 00 00 00 00 00 00 00  |dx..............|
> 000000a0  00 00 00 00 00 00 00 00  00 00 00 00              |............|
> 000000ac
> ```
> 
> So, I'll fix kernel client to send an old `struct ceph_mds_request_head` if MDS does not support `CEPHFS_FEATURE_HAS_OWNER_UIDGID`. It means that we need to update userspace client code too, and add some version checks to it to prevent any possible breakage.

Do you mean the old MDS also hasn't support the `CEPHFS_FEATURE_32BITS_RETRY_FWD` yet ?

mihalicyn commented:
> > It looks like even after [cbd7e30](https://github.com/ceph/ceph/commit/cbd7e3040208a5d89b9a95d4161dbd908da51079) we still need to handle each `CEPH_MDS_REQUEST_HEAD_VERSION` update separately. I've just tested old MDS (without extra fields in `struct ceph_mds_request_head`) with new kernel client and it fails:
> > ```
> > 2023-08-03T13:15:40.871+0200 7fe64ef5e640 10 mds.c ms_handle_accept v1:192.168.2.136:0/49354629 con 0x563962206880 session 0x563967054000
> > 2023-08-03T13:15:40.871+0200 7fe650f62640 -1 failed to decode message of type 24 v6: End of buffer [buffer:2]
> > 2023-08-03T13:15:40.871+0200 7fe650f62640  1 dump: 
> > 00000000  03 00 01 00 00 00 00 00  00 00 10 00 00 00 00 00  |................|
> > 00000010  00 00 00 00 00 00 01 01  00 00 00 00 00 00 00 00  |................|
> > 00000020  00 00 00 00 00 00 00 00  00 00 01 00 00 00 00 00  |................|
> > 00000030  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
> > *
> > 00000070  00 00 01 01 00 00 00 00  00 00 00 00 00 00 00 01  |................|
> > 00000080  00 00 00 00 00 00 00 00  00 00 00 00 5b 8c cb 64  |............[..d|
> > 00000090  64 78 11 13 01 00 00 00  00 00 00 00 00 00 00 00  |dx..............|
> > 000000a0  00 00 00 00 00 00 00 00  00 00 00 00              |............|
> > 000000ac
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So, I'll fix kernel client to send an old `struct ceph_mds_request_head` if MDS does not support `CEPHFS_FEATURE_HAS_OWNER_UIDGID`. It means that we need to update userspace client code too, and add some version checks to it to prevent any possible breakage.
> 
> Do you mean the old MDS also hasn't support the `CEPHFS_FEATURE_32BITS_RETRY_FWD` yet ?

No, "old MDS" was MDS without `CEPHFS_FEATURE_HAS_OWNER_UIDGID` and with `CEPHFS_FEATURE_32BITS_RETRY_FWD`.

You can easily reproduce this, just add a new field to the ceph_request_head structure in the kernel client and it will stop to work.
The problem is that MDS decoder is not ready to take `ceph_request_head` that bigger in size than expected. So, we need to always send `ceph_request_head` of an appropriate size.

mihalicyn commented:
@lxbsz let's continue our [discussion](https://lore.kernel.org/all/41ba61bf-3a45-cb20-1e4c-38dbd65bafa6@redhat.com/) about protocol change here.

I've this draft of implementation:
```
struct ceph_mds_request_head {
	__le16 version;
	__le64 oldest_client_tid;
	__le32 mdsmap_epoch;           /* on client */
	__le32 flags;                  /* CEPH_MDS_FLAG_* */
	__u8 num_retry, num_fwd;       /* legacy count retry and fwd attempts */
	__le16 num_releases;           /* # include cap/lease release records */
	__le32 op;                     /* mds op code */
	__le32 caller_uid, caller_gid;
	__le64 ino;                    /* use this ino for openc, mkdir, mknod,
					  etc. (if replaying) */
	union ceph_mds_request_args args;

	__le32 ext_num_retry;          /* new count retry attempts */
	__le32 ext_num_fwd;            /* new count fwd attempts */

	__le32 struct_len;             /* to store size of struct ceph_mds_request_head */
	__le32 owner_uid, owner_gid;   /* used for OPs which create inodes */
} __attribute__ ((packed));

...

void inline decode(struct ceph_mds_request_head& h, ceph::buffer::list::const_iterator& bl) {
  using ceph::decode;
  unsigned struct_end = bl.get_off();

  decode(h.version, bl);
  decode(h.oldest_client_tid, bl);
  decode(h.mdsmap_epoch, bl);
  decode(h.flags, bl);
  decode(h.num_retry, bl);
  decode(h.num_fwd, bl);
  decode(h.num_releases, bl);
  decode(h.op, bl);
  decode(h.caller_uid, bl);
  decode(h.caller_gid, bl);
  decode(h.ino, bl);
  bl.copy(sizeof(h.args), (char*)&(h.args));

  if (h.version >= 2) {
    decode(h.ext_num_retry, bl);
    decode(h.ext_num_fwd, bl);
  } else {
    h.ext_num_retry = h.num_retry;
    h.ext_num_fwd = h.num_fwd;
  }

  if (h.version >= 3) {
    decode(h.struct_len, bl);
    struct_end += h.struct_len;

    decode(h.owner_uid, bl);
    decode(h.owner_gid, bl);
  } else {
    /*
     * client is old: let's take caller_{u,g}id as owner_{u,g}id
     * this is how it worked before adding of owner_{u,g}id fields.
     */
    h.owner_uid = h.caller_uid;
    h.owner_gid = h.caller_gid;
  }

  /* add new fields handling here */

  /*
   * From version 3 we have struct_len field.
   * It allows us to properly handle a case
   * when client send struct ceph_mds_request_head
   * bigger in size than MDS supports. In this
   * case we just want to skip all remaining bytes
   * at the end.
   *
   * See also DECODE_FINISH macro. Unfortunately,
   * we can't start using it right now as it will be
   * an incompatible protocol change.
   */
  if (h.version >= 3) {
    if (bl.get_off() > struct_end)
      throw ::ceph::buffer::malformed_input(DECODE_ERR_PAST(__PRETTY_FUNCTION__));
    if (bl.get_off() < struct_end)
      bl += struct_end - bl.get_off();
  }
}
```

This is what you've proposed?

lxbsz commented:
> @lxbsz let's continue our [discussion](https://lore.kernel.org/all/41ba61bf-3a45-cb20-1e4c-38dbd65bafa6@redhat.com/) about protocol change here.
> 
[...]
> ```
> 
> This is what you've proposed?

Yeah, something likes this. And after this it will make the code simpler and easier to maintain after your change.

mihalicyn commented:
Ok, great. I'll test this change carefully and update this PR and kernel client.

Couldn't you check if there is anything else that should be fixed in the kernel client?

lxbsz commented:
> Ok, great. I'll test this change carefully and update this PR and kernel client.
> 
> Couldn't you check if there is anything else that should be fixed in the kernel client?

For now everything looks good to me, I will check more about that Monday.

mihalicyn commented:
New revision.

- this PR updated (added `struct_len` field as suggested by @lxbsz)
- kernel client updated https://lore.kernel.org/all/20230804084858.126104-1-aleksandr.mikhalitsyn@canonical.com/

lxbsz commented on src/include/ceph_fs.h:
```diff
@@ -430,6 +430,12 @@ enum {
 
+#define IS_CEPH_MDS_OP_NEWINODE(op) (op == CEPH_MDS_OP_CREATE     || \
+				     op == CEPH_MDS_OP_MKNOD      || \
+				     op == CEPH_MDS_OP_MKSNAP     || \
```
For `CEPH_MDS_OP_MKSNAP` is special and won't create new inode in `MDS`. We can just skip it.

mihalicyn replied:
yes, thanks!

lxbsz commented on src/mds/Server.cc:
```diff
@@ -3506,10 +3506,10 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
   }
 
-  _inode->uid = mdr->client_request->get_caller_uid();
+  _inode->uid = mdr->client_request->get_owner_uid();
```
Let's do a sanity check here to make sure the `owner_{uid/gid}` are correctly set. Which shouldn't be `-1`.

mihalicyn replied:
Fixed, thanks!

lxbsz commented on src/mds/Server.cc:
```diff
@@ -3506,10 +3506,14 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
-  _inode->uid = mdr->client_request->get_caller_uid();
+  _inode->uid = (mdr->client_request->get_owner_uid() != (unsigned)-1) ?
+		 mdr->client_request->get_owner_uid() :
+		 mdr->client_request->get_caller_uid();
```
Since you have force setting the `owner_{uid/gid}`, so we should `ceph_assert()` instead of setting them to `caller_{uid/gid}` here.

If they are still `-1` there should be buggy in MDS ?

mihalicyn replied:
I personally can't imagine how `-1` can appear there. I'm happy with any option, including `ceph_assert`.

Are you comfortable with `ceph_assert(_inode->uid != (unsigned)-1)`?

lxbsz replied:
In theory this should happen anyway, so this should be okay and could catch bugs.

mihalicyn replied:
> In theory this should happen anyway, so this should be okay and could catch bugs.

yes, I agree. So, I'm adding `ceph_assert(_inode->uid != (unsigned)-1)`, right? Alternatively, we can
add this check just in `get_owner_uid` implementation.

lxbsz replied:
> > In theory this should happen anyway, so this should be okay and could catch bugs.
> 
> yes, I agree. So, I'm adding `ceph_assert(_inode->uid != (unsigned)-1)`, right? Alternatively, we can add this check just in `get_owner_uid` implementation.

It will be more proper to check it here instead.

mihalicyn replied:
Good. Have fixed.

lxbsz reviewed (approved):
LGTM. Thanks @mihalicyn for your nice work.

vshankar commented:
jenkins retest this please

mihalicyn commented:
jenkins retest this please

vshankar commented:
https://pulpito.ceph.com/vshankar-2023-08-23_03:59:53-fs-wip-vshankar-testing-20230822.060131-testing-default-smithi/

(will review when the run finishes)

vshankar reviewed (approved):
https://tracker.ceph.com/projects/cephfs/wiki/Main#24-Aug-2023

mihalicyn commented:
Thanks a lot for your help with this @vshankar!

mihalicyn commented:
Speaking about backports. Do I need to prepare PRs for older releases by myself? Or this stuff is handled automatically somehow?

lxbsz commented:
> Speaking about backports. Do I need to prepare PRs for older releases by myself? Or this stuff is handled automatically somehow?

You need to backport it yourself one by one.

mihalicyn commented:
> > Speaking about backports. Do I need to prepare PRs for older releases by myself? Or this stuff is handled automatically somehow?
> 
> You need to backport it yourself one by one.

Thanks, Xiubo! Will do.
